ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/SourceBuilder.java

```diff
@@ -232,4 +266,116 @@ private static Formats of(final KeyFormat keyFormat, final FormatInfo valueForma
+        private ProcessorContext processorContext;
+
+        @Override
+        public void init(final ProcessorContext processorContext) {
+          this.processorContext = requireNonNull(processorContext, "processorContext");
+        }
+
+        @Override
+        public GenericRow transform(final K key, final GenericRow row) {
+          if (row == null) {
+            return row;
+          }
+
+          final int numPseudoColumns = SystemColumns
+              .pseudoColumnNames(pseudoColumnVersion).size();
+
+          row.ensureAdditionalCapacity(numPseudoColumns - 1);
+
+          if (pseudoColumnVersion >= SystemColumns.ROWPARTITION_ROWOFFSET_PSEUDOCOLUMN_VERSION) {
+            final int partition = processorContext.partition();
+            final long offset = processorContext.offset();
+            row.append(partition);
+            row.append(offset);
+          }
+
+          return row;
+        }
+
+        @Override
+        public void close() {
+        }
+      };
+    }
+  }
+
+  private static class AddRemainingPseudoAndKeyCols<K>
+      implements ValueTransformerWithKeySupplier<K, GenericRow, GenericRow> {
+
+    private final Function<K, Collection<?>> keyGenerator;
+    private final int pseudoColumnVersion;
+
+    AddRemainingPseudoAndKeyCols(
+        final Function<K, Collection<?>> keyGenerator, final int pseudoColumnVersion) {
+      this.keyGenerator = requireNonNull(keyGenerator, "keyGenerator");
+      this.pseudoColumnVersion = pseudoColumnVersion;
+    }
+
+    @Override
+    public ValueTransformerWithKey<K, GenericRow, GenericRow> get() {
+      return new ValueTransformerWithKey<K, GenericRow, GenericRow>() {
+        private ProcessorContext processorContext;
+
+        @Override
+        public void init(final ProcessorContext processorContext) {
+          this.processorContext = requireNonNull(processorContext, "processorContext");
+        }
+
+        @Override
+        public GenericRow transform(final K key, final GenericRow row) {
+          if (row == null) {
+            return row;
+          }
+
+          final Collection<?> keyColumns = keyGenerator.apply(key);
+
+          //remove pseudocolumns we previously materialized so we can add them back in correct order
+
+          //ensure extra capacity equal to number of pseudoColumns which we haven't materialized
+          final int pseudoColumnsToAdd = SystemColumns.pseudoColumnNames(
+              SystemColumns.ROWTIME_PSEUDOCOLUMN_VERSION).size();
+
+          row.ensureAdditionalCapacity(pseudoColumnsToAdd);
+
+          //calculate number of user columns and
+          final Set<ColumnName> columnNames = SystemColumns.pseudoColumnNames(pseudoColumnVersion);
+          final int totalPseudoColumns = columnNames.size();
+          final int pseudoColumnsToShift = totalPseudoColumns - pseudoColumnsToAdd;
+          final int numUserColumns = row.size() - pseudoColumnsToShift;
+
+          Object toShift = processorContext.timestamp();
```

Comment:
This pattern of shifting a single column works for now but does not extend nicely if we add more pseudocolumns in the future. I think it'd be cleaner to simply copy out the relevant pseudocolumns and then put them back in in the right places. That said, this can always be updated later so don't worry about it if you've got other things to work on.

Either way, can we add a comment to describe what this code is doing? Takes a while to decipher otherwise.

Reply:
I'll add a comment for now and a note to re-visit this.

Reply:
So I remember why I didn't make a solution with a list in the first place: I can't just use `appendAll()` after building the list, as `GenericRow` doesn't expose a `remove()` function or anything like that- also, since trying to use `set()` with an in-capacity but not yet populated index throws an out of bounds exception, I'd need to iterate an add the first n pseudocolumns, where n is the number of pseudocolumns which do need to be materialized and were therefore already in the row, by using `set()`, then appending the last ones. I think my method ends up being more readable than this, at least with a comment. Thoughts?

Reply:
How about calling `append()` the requisite number times to just append nulls (as a form of initialization), and then calling `set()` in a single for-loop?

Reply:
This would work well. My question is if the readability would be better at this point. At the least, I think it'll be easier to comment, and extend. I'll implement it and we can decide which way is better.